### PR TITLE
[hue] Decoupling of 'sensorRefreshInterval' and time gap for lastchange check

### LINKED ISSUE #6577: [hue] Unreliable Events from Hue Dimmer Switch
<a alt="bounty" href="https://www.bountysource.com/issues/85352560"></a>

Parallel to my “production” setup i got a docker container with the current 2.5 Milestone Build.

I found out that lights that i control with the Philips Hue Dimmer Switch via the Milestone Build i got unreliable events from the button.

Here is the events.log part of the Build:
```
2019-11-01 13:47:00.658 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 1002.0
2019-11-01 13:47:04.833 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 1002.0
2019-11-01 13:47:07.944 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 1002.0
2019-11-01 13:47:18.320 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 1002.0
2019-11-01 13:47:24.549 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 1000.0
2019-11-01 13:47:42.181 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 4000.0
2019-11-01 13:47:42.700 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 4002.0
2019-11-01 13:48:34.549 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 4000.0
2019-11-01 13:48:42.847 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 4002.0
2019-11-01 13:48:49.577 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 4000.0
2019-11-01 13:49:58.571 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 4002.0
2019-11-01 13:53:16.717 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 1002.0
2019-11-01 13:53:25.545 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:Zigbee_Kueche_Fernbedienung:dimmer_switch_event triggered 4000.0
```

As you can see the Short Release events is triggered 10 seconds after the initial pressed events. sometimes i got a lot of 4000.0 events and NO 4002.0 event.

In my production setup based on the 2.4 stable version that same event from the same bridge looks like this:

```
2019-11-01 13:47:00.496 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 1000.0
2019-11-01 13:47:01.225 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 1002.0
2019-11-01 13:47:04.707 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 1002.0
2019-11-01 13:47:07.827 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 1002.0
2019-11-01 13:47:15.110 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 1002.0
2019-11-01 13:47:18.254 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 1002.0
2019-11-01 13:47:24.483 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 1000.0
2019-11-01 13:47:25.005 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 1002.0
2019-11-01 13:47:37.986 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4002.0
2019-11-01 13:47:42.161 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4000.0
2019-11-01 13:47:42.722 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4002.0
2019-11-01 13:48:12.883 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4000.0
2019-11-01 13:48:13.405 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4002.0
2019-11-01 13:48:34.722 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4002.0
2019-11-01 13:48:42.519 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4000.0
2019-11-01 13:48:43.043 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4002.0
2019-11-01 13:48:49.817 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4002.0
2019-11-01 13:49:58.554 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4002.0
2019-11-01 13:53:16.897 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 1002.0
2019-11-01 13:53:25.755 [vent.ChannelTriggeredEvent] - hue:0820:00178827902c:kitchen:dimmer_switch_event triggered 4002.0
```
Basically i am missing a lot of *002.0 events or they got send much too late.

Should i open an issue on github for this?

The openhab.log file looks clean during this time period.

Here are some hints by cweitkamp of the root cause of this issue:
https://community.openhab.org/t/unreliable-philips-hue-dimmer-switch-events-with-current-2-5-milestone-m4-build/84650/19?u=regnets

### PR DESCRIPTION
**Pull request by cweitkamp**
- Decoupling of `sensorRefreshInterval` and time gap for lastchange check

The accuracy of time values returned by the hue API is seconds. The `sensorRefreshInterval` are milliseconds. Reducing the `sensorRefreshInterval` leads to ignored events because of that.

Fixes #6577
Does not fix https://github.com/openhab/openhab2-addons/issues/6577#issuecomment-565160525 or #6578

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/sensors/DimmerSwitchHandler.java [hunk 1]
```diff
@@ -96,7 +77,7 @@ protected void doSensorStateChanged(@Nullable HueBridge bridge, FullSensor senso
             String value = String.valueOf(buttonState);
             updateState(CHANNEL_DIMMER_SWITCH, new DecimalType(value));
             Instant then = timestamp.toInstant();
-            Instant someSecondsEarlier = now.minusNanos(refreshIntervalInNanos).toInstant();
+            Instant someSecondsEarlier = now.minusSeconds(5).toInstant();
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/sensors/DimmerSwitchHandler.java]: 5 seems like a random number - if I read the old code correctly, the intention was to use half of the refresh interval.

If you think 5 is always a good choice, you'd better define it as a constant with some comment next to it.

**Comment by cweitkamp** [comment]
If you do not link a channel #6578 will not affect you and you can test with a normal / low cpu load.

The hue API issues one event per second (at maximum) and their timestamp looks like this: `"lastupdated": "yyyy-mm-ddThh:mm:ss"`. The `sensorRefreshInterval` is given in milliseconds. Before triggering an event we check if the `"lastupdated"` timestamp is between `now` and `now` minus twice the `sensorRefreshInterval`.

```
now - 2 * sensorRefreshInterval < lastupdated < now
```

This could lead to the following case when that skips each event:

"lastupdated" = 2019-12-13T16:00:30
"now" = 2019-12-13T16:00:30.500
"sensorRefreshInterval" = 50ms (the minimum allowed value)
"now - 2 * sensorRefreshInterval" = 2019-12-13T16:00:30.500 - 2 * 50ms = 2019-12-13T16:00:30.400

=> event will be skipped.

This will not happen anymore when we increase the value to check against. `now` minus 5s is a safe value. I think 2s or (3s will work too).

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/sensors/DimmerSwitchHandler.java (hunk 1 above)
Comment (cweitkamp) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/sensors/DimmerSwitchHandler.java]: I added a comment inside the code.

**Comment by kaikreuzer** [comment]
Thanks for the clarification.

One last question: Should maybe the public method `HueBridgeHandler.getSensorPollingInterval()` be removed, now that it isn't used anymore by anybody?

**Comment by cweitkamp** [comment]
> Should maybe the public method `HueBridgeHandler.getSensorPollingInterval()` be removed

Yes, fair enough. Done.

**Comment by kaikreuzer** [comment]
Can we ignore the failing hue tests on Travis?

**Comment by cweitkamp** [comment]
It builds locally. Looks like similar hick-ups we saw in the TRADFRI PRs.

**Comment by kaikreuzer** [comment]
Yep, that's what I thought as well.

**Comment by RobertOnGit** [comment]
Today I updated my openHAB system to 2.5.1. The first confrontation with the Dimmer Switch regression is rather dramatic (OH2.4 worked perfectly fine). Now at least 75% of the events are missed, rendering the Dimmer Switches more or less unusable...

I've done a brief test (for multiple RCs and all of the buttons): single press, several times a single press, press and hold once, press and hold at least 2 repeats.

So I can confirm the regression findings reported by others.

**Pull request merged**